In [61]:
import pandas as pd

In [62]:
data = pd.read_csv("titanic_train.csv")
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

# Cleaning the Data

### As name, ticket, cabin, PassengerId wouldnt help us so dropping them but first making a copy of the data

In [63]:
data_copy = data.copy()
data_copy.drop(["Name", "Ticket", "PassengerId", "Cabin"], axis=1, inplace=True)
data_copy.head()

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    male  22.0      1      0   7.2500        S
1         1       1  female  38.0      1      0  71.2833        C
2         1       3  female  26.0      0      0   7.9250        S
3         1       1  female  35.0      1      0  53.1000        S
4         0       3    male  35.0      0      0   8.0500        S

In [64]:
data_copy.shape

(891, 8)

In [65]:
data_copy.describe()

Survived      Pclass         Age       SibSp       Parch        Fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

## Now for NaN Values we can clearly see that count of Age is less than 891

In [66]:
data_copy["Age"].fillna(data["Age"].median(), inplace=True)
data_copy.describe()

Survived      Pclass         Age       SibSp       Parch        Fare
count  891.000000  891.000000  891.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.361582    0.523008    0.381594   32.204208
std      0.486592    0.836071   13.019697    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   22.000000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   35.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

## Cols contaning string arent in describe section so filling them independently

In [67]:
data_copy["Embarked"].fillna("U", inplace=True)
data_copy.head()

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    male  22.0      1      0   7.2500        S
1         1       1  female  38.0      1      0  71.2833        C
2         1       3  female  26.0      0      0   7.9250        S
3         1       1  female  35.0      1      0  53.1000        S
4         0       3    male  35.0      0      0   8.0500        S

# As we know we cant process further with strings so converting them using mapping
### Its basically denoting 0 for male and 1 for female and same going 1,2,3,4 for C,Q,S,U like that

In [68]:
from sklearn import preprocessing as pp

In [69]:
genders = {"male" : 0, "female" : 1}
data1 = [data_copy]

embarked = {"U" : 0, "S" : 1, "Q" : 2, "C" : 3}


for datasets in data1 :
    datasets['Sex'] = datasets['Sex'].map(genders)
    datasets['Embarked'] = datasets['Embarked'].map(embarked)

data_copy

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked
0           0       3    0  22.0      1      0   7.2500         1
1           1       1    1  38.0      1      0  71.2833         3
2           1       3    1  26.0      0      0   7.9250         1
3           1       1    1  35.0      1      0  53.1000         1
4           0       3    0  35.0      0      0   8.0500         1
..        ...     ...  ...   ...    ...    ...      ...       ...
886         0       2    0  27.0      0      0  13.0000         1
887         1       1    1  19.0      0      0  30.0000         1
888         0       3    1  28.0      1      2  23.4500         1
889         1       1    0  26.0      0      0  30.0000         3
890         0       3    0  32.0      0      0   7.7500         2

[891 rows x 8 columns]

# Now our training data is cleaned and now we can go for model implementation

In [70]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(max_iter=1000)

In [71]:
# Getting x_train and y_train
x_train = data_copy.iloc[:,1:]
y_train = data_copy.iloc[:,0]

In [72]:
x_train

Pclass  Sex   Age  SibSp  Parch     Fare  Embarked
0         3    0  22.0      1      0   7.2500         1
1         1    1  38.0      1      0  71.2833         3
2         3    1  26.0      0      0   7.9250         1
3         1    1  35.0      1      0  53.1000         1
4         3    0  35.0      0      0   8.0500         1
..      ...  ...   ...    ...    ...      ...       ...
886       2    0  27.0      0      0  13.0000         1
887       1    1  19.0      0      0  30.0000         1
888       3    1  28.0      1      2  23.4500         1
889       1    0  26.0      0      0  30.0000         3
890       3    0  32.0      0      0   7.7500         2

[891 rows x 7 columns]

In [73]:
y_train

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    0
889    1
890    0
Name: Survived, Length: 891, dtype: int64

In [74]:
model.fit(x_train,y_train)

LogisticRegression(max_iter=1000)

# We have to clean the test data too so going for that now

In [75]:
data_test = pd.read_csv("titanic_test.csv")
data_test_copy = data_test.copy()
data_test_copy.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [76]:
y_res = data_test_copy["PassengerId"]
y_res

0       892
1       893
2       894
3       895
4       896
       ... 
413    1305
414    1306
415    1307
416    1308
417    1309
Name: PassengerId, Length: 418, dtype: int64

In [77]:
data_test_copy.drop(["Name", "Ticket", "PassengerId", "Cabin"], axis=1, inplace=True)
data_test_copy.head()

Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0       3    male  34.5      0      0   7.8292        Q
1       3  female  47.0      1      0   7.0000        S
2       2    male  62.0      0      0   9.6875        Q
3       3    male  27.0      0      0   8.6625        S
4       3  female  22.0      1      1  12.2875        S

In [78]:
data_test_copy.describe()

Pclass         Age       SibSp       Parch        Fare
count  418.000000  332.000000  418.000000  418.000000  417.000000
mean     2.265550   30.272590    0.447368    0.392344   35.627188
std      0.841838   14.181209    0.896760    0.981429   55.907576
min      1.000000    0.170000    0.000000    0.000000    0.000000
25%      1.000000   21.000000    0.000000    0.000000    7.895800
50%      3.000000   27.000000    0.000000    0.000000   14.454200
75%      3.000000   39.000000    1.000000    0.000000   31.500000
max      3.000000   76.000000    8.000000    9.000000  512.329200

In [79]:
data_test_copy['Age'].fillna(data_test_copy['Age'].median(), inplace=True)
data_test_copy['Fare'].fillna(data_test_copy['Fare'].mean(), inplace=True)
data_test_copy.describe()

Pclass         Age       SibSp       Parch        Fare
count  418.000000  418.000000  418.000000  418.000000  418.000000
mean     2.265550   29.599282    0.447368    0.392344   35.627188
std      0.841838   12.703770    0.896760    0.981429   55.840500
min      1.000000    0.170000    0.000000    0.000000    0.000000
25%      1.000000   23.000000    0.000000    0.000000    7.895800
50%      3.000000   27.000000    0.000000    0.000000   14.454200
75%      3.000000   35.750000    1.000000    0.000000   31.500000
max      3.000000   76.000000    8.000000    9.000000  512.329200

In [85]:
data_test_copy['Embarked'].fillna("U", inplace=True)
data_test_copy.head()

Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0       3    male  34.5      0      0   7.8292        Q
1       3  female  47.0      1      0   7.0000        S
2       2    male  62.0      0      0   9.6875        Q
3       3    male  27.0      0      0   8.6625        S
4       3  female  22.0      1      1  12.2875        S

In [87]:
# mapping the values for the test data
genders_test = {"male" : 0, "female" : 1}
embarked_test = {"U" : 0, "S" : 1, "Q" : 2, "C" : 3}
data11 = [data_test_copy]
for i in data11:
    i['Sex'] = i['Sex'].map(genders_test)
    i['Embarked'] = i['Embarked'].map(embarked_test)

data_test_copy.head()

Pclass  Sex   Age  SibSp  Parch     Fare  Embarked
0       3    0  34.5      0      0   7.8292         2
1       3    1  47.0      1      0   7.0000         1
2       2    0  62.0      0      0   9.6875         2
3       3    0  27.0      0      0   8.6625         1
4       3    1  22.0      1      1  12.2875         1

In [90]:
y_pred = model.predict(data_test_copy)
y_pred

array([0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1,
       1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1,
       1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1,
       1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1,
       0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0,
       1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1,
       0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0,
       0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0,

In [94]:
df = pd.DataFrame({"PassengerId" : y_res.values,
                   "Survived" : y_pred})
df

PassengerId  Survived
0            892         0
1            893         0
2            894         0
3            895         0
4            896         1
..           ...       ...
413         1305         0
414         1306         1
415         1307         0
416         1308         0
417         1309         0

[418 rows x 2 columns]

In [95]:
df.to_csv("Submission.csv", index=False)